embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceCacheService.kt

```diff
@@ -57,13 +59,27 @@ internal class EmbraceCacheService(
         }
     }
 
+    /**
+     * Provides a function that writes the bytes from a cached file, if it exists, to an
+     * output stream and compresses it if it's not already compressed.
+     * We are storing compressed files from version 6.3.0 onwards. In order to avoid loosing
+     * payloads from previous versions, we need to check if the file is compressed or not.
+     */
     override fun loadPayload(name: String): SerializationAction {
         logger.logDeveloper(TAG, "Attempting to read bytes from $name")
         return { stream ->
             val file = storageService.getFileForRead(EMBRACE_PREFIX + name)
             try {
-                file.inputStream().buffered().use { input ->
-                    input.copyTo(stream)
+                if (compressor.isCompressed(file)) {
+                    file.inputStream().buffered().use { input ->
+                        input.copyTo(stream)
+                    }
+                } else {
+                    compressor.compress(stream) {
+                        file.inputStream().buffered().use { input ->
+                            input.copyTo(it)
+                        }
+                    }
```

Comment:
Could `Compressor` extend `OutputStream` and delegate most of its work (apart from checking whether info is compressed) to `GzipOutputStream`? IMO this would avoid having to introduce logic like this throughout the codebase